### PR TITLE
Add basic auth to artifacts request

### PR DESCRIPTION
**Pull request by venigreat**
Add a new dictionary to buildArtifacts
for example "hostsTokens": {"host": "token"}

Dictionary values ​​are added to artifact requests if they need authorization

### CODE REVIEW

File: Sources/BuildArtifacts/BuildArtifacts.swift [hunk 1]
```diff
@@ -12,24 +12,30 @@ public struct BuildArtifacts: Codable, Hashable, CustomStringConvertible {
     
     /// Location of additional apps that can be launched diring tests.
     public let additionalApplicationBundles: [AdditionalAppBundleLocation]
+    
+    /// Dictionary with Host : Token for artifacts urls
+    public let hostsTokens: [String: String]
```
Comment (beefon) [inline Sources/BuildArtifacts/BuildArtifacts.swift:17]: Let's extend this feature a bit. Can you please rename this to `hostHeaders: [String: [String: String]]` to allow appending an arbitrary HTTP headers to outgoing requests.

**Review by beefon — CHANGES_REQUESTED** [review]
Thanks for doing this!

**Comment by beefon** [comment]
You also need to attach your headers to `ResourceLocation.remoteUrl`, otherwise headers won't be applied on workers.

**Comment by beefon** [comment]
Want to clarify `ResourceLocation`. Emcee uses 'typed' resource locations to make sure you can't pass a [location of an app](https://github.com/avito-tech/Emcee/blob/master/Sources/BuildArtifacts/AppBundleLocation.swift) instead of [xctest bundle location](https://github.com/avito-tech/Emcee/blob/master/Sources/BuildArtifacts/TestBundleLocation.swift), for example. There are other types of resource locations (just a wrappers).

So by saying..

> You also need to attach your headers to `ResourceLocation.remoteUrl`

I meant updating a JSON scheme to add headers to `ResourceLocation.remoteUrl(URL, headers)` case.

Thus, instead of adding headers to [build artifacts](https://github.com/avito-tech/Emcee/blob/master/Sources/BuildArtifacts/BuildArtifacts.swift) model, it is better to add them to [resource location](https://github.com/avito-tech/Emcee/blob/master/Sources/ResourceLocation/ResourceLocation.swift), because individual artifacts potentially may be hosted on different hosts, and Emcee adopters might want to send different headers for different kinds of artifacts. Potentially.

**Comment by beefon** [comment]
Want to clarify `ResourceLocation`. Emcee uses 'typed' resource locations to make sure you can't pass a [location of an app](https://github.com/avito-tech/Emcee/blob/master/Sources/BuildArtifacts/AppBundleLocation.swift) instead of [xctest bundle location](https://github.com/avito-tech/Emcee/blob/master/Sources/BuildArtifacts/TestBundleLocation.swift), for example. There are other types of resource locations (just a wrappers).

So by saying..

> You also need to attach your headers to `ResourceLocation.remoteUrl`

I meant updating a JSON scheme to add headers to `ResourceLocation.remoteUrl(URL, headers)` case.

Thus, instead of adding headers to [build artifacts](https://github.com/avito-tech/Emcee/blob/master/Sources/BuildArtifacts/BuildArtifacts.swift) model, it is better to add them to [resource location](https://github.com/avito-tech/Emcee/blob/master/Sources/ResourceLocation/ResourceLocation.swift), because individual artifacts potentially may be hosted on different hosts, and Emcee adopters might want to send different headers for different kinds of artifacts. Potentially.

**Comment by venigreat** [comment]
All comments fixed, all tests is green

File: Tests/TestArgFileTests/TestArgFileTests.swift [hunk 2]
```diff
@@ -118,13 +118,24 @@ final class TestArgFileTests: XCTestCase {
...
+                            "runtime": "11.3"
+                        },
                         "testType": "uiTest",
                         "buildArtifacts": {
-                            "appBundle": "http://example.com/App.zip#MyApp/MyApp.app",
-                            "runner": "http://example.com/App.zip#Tests/UITests-Runner.app",
-                            "xcTestBundle": "http://example.com/App.zip#Tests/UITests-Runner.app/PlugIns/UITests.xctest"
+                            "appBundle": {
+                                "url": "http://example.com/App.zip#MyApp/MyApp.app"
+                            },
```
Comment (beefon) [inline Tests/TestArgFileTests/TestArgFileTests.swift:132]: Do you think you can update your PR to support old styled URLs for cases when there are no headers needed?

```
"appBundle": {
    "url": "http://example.com/App.zip#MyApp/MyApp.app",
    "headers": {"x": "y"}
}
```

```
"appBundle": "http://example.com/App.zip#MyApp/MyApp.app"
```

File: Sources/ResourceLocation/ResourceLocation.swift [hunk 3]
```diff
@@ -26,20 +26,34 @@ public enum ResourceLocation: Hashable, CustomStringConvertible, Codable {
...
+            return nil
+        }
+    }
+    
+    private enum CodingKeys: CodingKey {
+        case url
+        case headers
+    }
+    
+    public static func from(_ string: String, headers: [String: String] = [:]) throws -> ResourceLocation {
```
Comment (beefon) [inline Sources/ResourceLocation/ResourceLocation.swift:50]: I will think about how we can make this function a bit cleaner after you merge your changes. For now let's leave it as is.

**Review by beefon — APPROVED** [review]

**Comment by beefon** [comment]
If you want, you can put up another PR with my minor comments.

**Pull request merged**